### PR TITLE
Correctly handle ops on optree node boundaries

### PR DESCRIPTION
**Pull request by alexjg**
The `SeekOp` query can produce incorrect results when the optree it is searching only has visible ops on the internal nodes. The issue is that in the `SeekOp` we skip an entire child node if it only contains invisible ops. Consider this optree:

seekop

In this optree the only visible ops for the "a" key and "b" key are in the non-leaf node. This means that when we query for the insert position for a new op on the "a" key we seek past the first node, but then because the next child node does not contain any visible ops either, we seek past it to the next child, even though none of the ops in it are for "a", consequently we end up calculating that the insert index should be 31,  rather than 16.

This problem has raised it's head when loading saved documents which were generated from such situations. This is because the new op ends up inserted out of order and we assume that incoming ops are in document order when loading them - this invariant being broken means we are not able to correctly calculate predecessors for the new ops and consequently we end up with mismatched heads errors.

I've "fixed" this by not skipping invisible ops. This seems like it would have quite a severe performance penalty and is just unsatisfying so we should probably do something else. @orionz this code is more your wheelhouse than mine, any thoughts?

### CODE REVIEW

**Comment by jeffa5** [comment]
When we find an empty (in terms of visibility) node would it be enough to check if the first op has the same key? If it is the same then we can assume there is still another visible element in another later node so skipping is ok. If the key is different then we know that the position we just found was the last one and so we're done?

**Comment by alexjg** [comment]
That's sort of the logic we want, the problem though is that at the point we're implementing the skip logic we're looking at an internal node of the tree so I think we need to add some additional functionality to the `TreeQuery` trait to allow "peeking" at the next op.

To elaborate, here's the problematic code from `SeekOp`.

```rust
impl<'a> TreeQuery<'a> for SeekOp<'a> {
    fn query_node_with_metadata(
        &mut self,
        child: &OpTreeNode,
        m: &OpSetMetadata,
        ops: &[Op],
    ) -> QueryResult {
        ...
        match self.op.key {
            ...
            Key::Map(_) => {
                if let Some(start) = self.start {
                    if self.pos + child.len() >= start {

                        // skip empty nodes
                        if child.index.visible_len(ListEncoding::List) == 0 { // <--- This is the problematic part
                            self.pos += child.len();
                            QueryResult::Next
                        } else {
                            QueryResult::Descend
                        }


                    } else {
                        self.pos += child.len();
                        QueryResult::Next
                    }
                } else {
                    ...
                }
            }
        }
    }

```

At the point that we decide to skip the invisible ops we don't have any way of checking whether the next op is for the same key because `child` could also be an internal node, we need to recursively descend into the next node until we find a leaf. We _could_ implement that logic in the query, but I wonder if it makes more sense to be part of the `search` infrastructure?

**Comment by alexjg** [comment]
Actually wait, I think you're idea is simpler to implement than I was imagining, it might do the trick. I'm not at a computer right now but will give it a shot when I'm back.

**Comment by alexjg** [comment]
Okay, @jeffa5 thanks for the suggestion, I've implemented it, the tests pass and I think it's sensible.

**Comment by jeffa5** [comment]
> Okay, @jeffa5 thanks for the suggestion, I've implemented it, the tests pass and I think it's sensible.

Thanks! I think it looks good, just left a small comment.

File: rust/automerge/src/query/seek_op.rs [hunk 1]
```diff
@@ -76,8 +76,19 @@ impl<'a> TreeQuery<'a> for SeekOp<'a> {
                     if self.pos + child.len() >= start {
                         // skip empty nodes
                         if child.index.visible_len(ListEncoding::List) == 0 {
-                            self.pos += child.len();
-                            QueryResult::Next
+                            let child_contains_key =
+                                child.elements.iter().any(|e| ops[*e].key == self.op.key);
```
[the remark this answers is not in the document]
Comment (alexjg) [reply rust/automerge/src/query/seek_op.rs:80]: I'm worried about the situation where you `start` at some offset in a non-leaf node. In this case `query_node_with_metadata` would get called with a node where the first element is _not_ for the same key. But the operations for the previous key could all be deleted, so the block still reports itself as invisible and if we only check the first element we wouldn't notice.

**Pull request merged**